client/packages/coldchain/src/Equipment/ImportAsset/UploadTab.tsx

```diff
@@ -30,21 +39,96 @@ interface ParsedAsset {
   [key: string]: string | undefined;
 }
 
-enum AssetColumn {
-  ASSET_NUMBER = 0,
-  CATALOGUE_ITEM_CODE = 1,
-  NOTES = 2,
-  SERIAL_NUMBER = 3,
-  INSTALLATION_DATE = 4,
-}
+const formatDate = (value: string): string | null =>
+  Formatter.naiveDate(DateUtils.getDateOrNull(value));
 
-// the row object indexes are returned in column order
-// which allows us to index the keys
-const getCell = (row: ParsedAsset, index: AssetColumn) => {
-  const rowKeys = Object.keys(row);
-  const key = rowKeys[index] ?? '';
-  return row[key] ?? '';
-};
+function getImportHelpers<T, P>(
+  row: P,
+  rows: T[],
+  index: number,
+  t: TypedTFunction<LocaleKey>
+) {
+  const importRow = {
+    id: FnUtils.generateUUID(),
+  } as T;
+  const rowErrors: string[] = [];
+
+  const addCell = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+    if (value !== undefined) {
+      (importRow[key] as unknown) = formatter
+        ? formatter(value as string)
+        : value;
+    }
+  };
+
+  const addRequired = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+
+    if (value === undefined || (value as string).trim() === '') {
+      rowErrors.push(
+        t('error.field-must-be-specified', {
+          field: t(localeKey),
+        })
+      );
+      return;
+    }
+
+    addCell(key, localeKey, formatter);
+  };
+
+  const addUnique = (
+    key: keyof T,
+    localeKey: LocaleKey,
+    formatter?: (value: string) => unknown
+  ) => {
+    const prop = t(localeKey) as keyof P;
+    const value = row[prop] ?? '';
+
+    addRequired(key, localeKey, formatter);
+
+    // check for duplicates
+    if (rows.some((r, i) => r[key] === value && index !== i)) {
+      rowErrors.push(t('error.duplicate-asset-number'));
```

Comment:
I wonder if this should be generic like the above error?

Something like
```
t('error.duplicate-field-error', {field: t(localeKey), })
```

Reply:
But probably not required until it is moved to some utils file

Reply:
nice! thanks - will pop that in